kairos-server/src/deposit_manager/mod.rs

```diff
@@ -0,0 +1,191 @@
+        .await?;
+
+        tracing::info!("Schemas fetched successfully");
+
+        let last_processed_deposit_index: u32 =
+            get_last_deposit_counter(casper_rpc_url, contract_hash)
+                .await
+                .map_err(|err| {
+                    DepositManagerError::UnexpectedError(format!(
+                        "Failed to fetch the index for the last processed deposit: {}",
+                        err
+                    ))
+                })?;
+
+        tracing::info!("On-chain stored last processed deposit index fetched successfully");
+
+        Ok(DepositManager {
+            last_deposit_added_to_batch: AtomicU32::new(last_processed_deposit_index + 1),
+            fetcher,
+            schemas,
+            known_deposit_deploys: RwLock::new(HashSet::new()),
+        })
+    }
+
+    /// Processes new events starting from the last known event ID.
+    pub async fn add_new_events_to(
+        &self,
+        batch_state_manager: &BatchStateManager,
+    ) -> Result<(), DepositManagerError> {
+        let last_unprocessed_deposit_index = self.fetcher.fetch_events_count().await?;
+
+        for deposit_index in (self
+            .last_deposit_added_to_batch
+            .fetch_add(1, Ordering::SeqCst)
+            + 1)..=last_unprocessed_deposit_index
+        {
+            let untyped_event = self
+                .fetcher
+                .fetch_event(deposit_index, &self.schemas)
+                .await?;
+
+            // (koxu1996) NOTE: I think we should rather use full transaction data (ASN) for events,
+            // parse them here with `kairos-tx` and then push to Data Availability layer.
+
+            match untyped_event.name.as_str() {
+                "Deposit" => {
+                    let data = untyped_event.to_ces_bytes()?;
+                    let (deposit, _) =
+                        contract_utils::Deposit::from_bytes(&data).map_err(|err| {
+                            DepositManagerError::UnexpectedError(format!(
+                                "Failed to parse deposit event from bytes: {}",
+                                err
+                            ))
+                        })?;
+                    batch_state_manager
+                        .enqueue_transaction(KairosTransaction::Deposit(L1Deposit {
+                            recipient: "cafebabe".into(),
+                            amount: deposit.amount,
+                        }))
+                        .await
+                        .map_err(|err| {
+                            DepositManagerError::UnexpectedError(format!(
+                                "Failed to parse deposit event from bytes: {}",
+                                err
+                            ))
+                        })?;
+                }
+                other => {
+                    tracing::error!("Unknown event type: {}", other)
+                }
+            }
+        }
+        self.last_deposit_added_to_batch
```

Comment:
Adding events to the batch state manager should be/ is invariant according to @Avi-D-coder. I may understood him wrong. If it's not I thing the better API is to make adding deposits to the batch_state_manager invariant in the batch_state_manager instead of making that a concern of the user of the API.

Reply:
Duplicate deposits added to a batch are not currently handled by the batch state manager. I could de-duplicate them easily.

Reply:
Well, instead trying to deduplicate deposits, we should not approve "refactor" code that introduces them.